File: go/coptic_date.go
package kiahk

import "fmt"

// CopticDate is a calendar date on the Coptic (Anno Martyrum) calendar.
type CopticDate struct {
	Year, Month, Day int
}

// isCopticLeap reports whether the given Coptic year is a leap year
// (Julian-style rule: Y mod 4 == 3).
func isCopticLeap(y int) bool { return y%4 == 3 }

// NewCopticDate validates the inputs and returns a CopticDate.
// On invalid input it returns the zero value and a *InvalidCopticDateError.
func NewCopticDate(year, month, day int) (CopticDate, error) {
	if month < 1 || month > 13 {
		return CopticDate{}, &InvalidCopticDateError{
			Year: year, Month: month, Day: day,
			Reason: fmt.Sprintf("coptic month must be 1..13, got %d", month),
		}
	}
	maxDay := 30
	if month == 13 {
		maxDay = 5
		if isCopticLeap(year) {
			maxDay = 6
		}
	}
	if day < 1 || day > maxDay {
		return CopticDate{}, &InvalidCopticDateError{
			Year: year, Month: month, Day: day,
			Reason: fmt.Sprintf("coptic day must be 1..%d for year %d month %d", maxDay, year, month),
		}
	}
	return CopticDate{Year: year, Month: month, Day: day}, nil
}

// ToGregorian converts this Coptic date to a Gregorian date.
// The constructor cannot reject the produced gregorian date; the error is discarded.
func (c CopticDate) ToGregorian() GregorianDate {
	y, m, d := CopticToGregorian(c.Year, c.Month, c.Day)
	g, _ := NewGregorianDate(y, m, d)
	return g
}

File: c/src/coptic_date.c
#include "kiahk.h"

static int is_coptic_leap(int y) { return y % 4 == 3; }

kiahk_error kiahk_coptic_date_init(kiahk_coptic_date *out, int year, int month, int day) {
    if (!out) return KIAHK_ERR_INVALID_COPTIC_DATE;
    if (month < 1 || month > 13) {
        out->year = 0; out->month = 0; out->day = 0;
        return KIAHK_ERR_INVALID_COPTIC_DATE;
    }
    int max_day = 30;
    if (month == 13) {
        max_day = is_coptic_leap(year) ? 6 : 5;
    }
    if (day < 1 || day > max_day) {
        out->year = 0; out->month = 0; out->day = 0;
        return KIAHK_ERR_INVALID_COPTIC_DATE;
    }
    out->year = year; out->month = month; out->day = day;
    return KIAHK_OK;
}

kiahk_error kiahk_coptic_date_to_gregorian(const kiahk_coptic_date *c, kiahk_gregorian_date *out) {
    if (!c || !out) return KIAHK_ERR_INVALID_COPTIC_DATE;
    int gy, gm, gd;
    kiahk_coptic_to_gregorian(c->year, c->month, c->day, &gy, &gm, &gd);
    return kiahk_gregorian_date_init(out, gy, gm, gd);
}
